llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp

```diff
@@ -1341,6 +1342,71 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
       Call->takeName(&II);
       return IC.replaceInstUsesWith(II, Call);
     }
+
+    // Fold ballot intrinsic based on llvm.assume hint about the result.
+    //
+    // assume(ballot(x) == ballot(true)) -> x = true
+    // assume(ballot(x) == -1)           -> x = true
+    // assume(ballot(x) == 0)            -> x = false
```

Comment:
We allow ballot.i32 in wave64 mode, which would break these optimizations.

Reply:
If you are going to all that trouble, then the tests should check all four combinations: i32 ballot in wave32, i32 ballot in wave64, etc.

Reply:
I covered these 4 combinations in the existing test cases - 
- 13 cases in each of `ballot.i32` on `wave32` & `ballot.i64` on `wave64`
- 2 base cases in each of the remaining 2 combinations